src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs

```diff
@@ -36,7 +34,40 @@ internal class SourceLocalSymbol : LocalSymbol
         private readonly LocalDeclarationKind _declarationKind;
         private readonly ScopedKind _scope;
 
-        private TypeWithAnnotations.Boxed _type;
+#nullable enable
+
+        private TypeWithAnnotations.Boxed? _type;
+
+        // Please don't use thread local storage widely. This should be one of only a few uses.
+        [ThreadStatic] private static PooledHashSet<LocalTypeInferenceInProgressKey>? s_LocalTypeInferenceInProgress;
```

Comment:
I didn't understand why we need two data structures and why we need to track syntax nodes (`LocalTypeInferenceInProgressKey.Reference` and `_forbiddenReferences`). 
Why isn't tracking the set of "locals being bound" (a `PooledHashSet<SourceLocalSymbol>` or `ConsList<SourceLocalSymbol>`) and saving the error type into `_type` sufficient to break the cycles?